source/common/filesystem/posix/directory_iterator_impl.cc

```diff
@@ -0,0 +1,61 @@
+#include "envoy/common/exception.h"
+
+#include "common/common/fmt.h"
+#include "common/filesystem/directory_iterator_impl.h"
+
+namespace Envoy {
+namespace Filesystem {
+
+DirectoryIteratorImpl::~DirectoryIteratorImpl() {
+  if (dir_ != nullptr) {
+    ::closedir(dir_);
+  }
+}
+
+DirectoryIterator::DirectoryEntry DirectoryIteratorImpl::nextEntry() {
+  if (dir_ == nullptr) {
+    openDirectory();
+  }
+
+  errno = 0;
+  dirent* entry = ::readdir(dir_);
+  if (entry == nullptr && errno != 0) {
+    throw EnvoyException(fmt::format("unable to iterate directory: {}", directory_path_));
+  }
+
+  if (entry == nullptr) {
+    return {"", FileType::Other};
+  }
+
+  const std::string current_path = std::string(entry->d_name);
+  const std::string full_path = directory_path_ + "/" + current_path;
+  return {current_path, fileType(full_path)};
```

Comment:
Seems kind of expensive to both copy and return the full path?

Reply:
It doesn't look like we copy or return the full path? The fileType function takes it by reference and returns a `DirectoryIterator::FileType`